components/forms/CreateTxForm/index.tsx

```diff
@@ -2,47 +2,59 @@ import { Account, calculateFee } from "@cosmjs/stargate";
 import { assert } from "@cosmjs/utils";
 import axios from "axios";
 import { NextRouter, withRouter } from "next/router";
-import { useState } from "react";
+import { useRef, useState } from "react";
 import { useAppContext } from "../../../context/AppContext";
-import { capitalizeFirstLetter } from "../../../lib/displayHelpers";
-import { TxMsg, TxType } from "../../../types/txMsg";
+import { gasOfTx } from "../../../lib/txMsgHelpers";
+import { MsgType, TxMsg } from "../../../types/txMsg";
 import Button from "../../inputs/Button";
 import Input from "../../inputs/Input";
 import StackableContainer from "../../layout/StackableContainer";
 import MsgForm from "./MsgForm";
 
+export interface MsgGetter {
+  readonly isMsgValid: (msg: TxMsg) => msg is TxMsg;
+  readonly msg: TxMsg;
+}
+
+const getMsgFormKey = (msgType: MsgType, msg: TxMsg) => JSON.stringify({ msgType, msg });
+
 interface CreateTxFormProps {
   readonly router: NextRouter;
-  readonly txType: TxType;
   readonly senderAddress: string;
   readonly accountOnChain: Account;
-  readonly closeForm: () => void;
 }
 
-const CreateTxForm = ({
-  router,
-  txType,
-  senderAddress,
-  accountOnChain,
-  closeForm,
-}: CreateTxFormProps) => {
+const CreateTxForm = ({ router, senderAddress, accountOnChain }: CreateTxFormProps) => {
   const { state } = useAppContext();
 
   const [processing, setProcessing] = useState(false);
-  const [checkAndGetMsg, setCheckAndGetMsg] = useState<() => TxMsg | null>();
+  const [msgTypes, setMsgTypes] = useState<readonly MsgType[]>([]);
+  const msgGetters = useRef<MsgGetter[]>([]);
   const [memo, setMemo] = useState("");
-  const [gasLimit, setGasLimit] = useState(200000);
+  const [gasLimit, setGasLimit] = useState(100_000);
```

Comment:
Can we set this to NaN or 0 and ensure the actual value is always calculated?

Reply:
Maybe a better representation of the initial state would be `useState(gasOfTx([]))`?. That way it'd be calculated and we'd have a single source of truth for the signature verification gas limit.

Reply:
Yeah, sounds very clean!